pennylane_sf/gbs.py

```diff
@@ -0,0 +1,246 @@
+    def apply(self, operation, wires, par):
+        self._params, A, _ = par
+
+        if len(self._params) != self.num_wires:
+            raise ValueError(
+                "The number of variable parameters must be equal to the total number " "of wires."
+            )
+
+        self._WAW = self._calculate_WAW(*par)
+        n_mean_WAW = self._calculate_n_mean(self._WAW)
+
+        op = self._operation_map[operation](self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
+        op | [self.q[wires.index(i)] for i in wires]
+
+        if not self.analytic:
+            MeasureFock() | [self.q[wires.index(i)] for i in wires]
+
+    def pre_measure(self):
+        self.eng = sf.Engine(self.backend, backend_options={"cutoff_dim": self.cutoff})
+
+        if self.analytic:
+            results = self.eng.run(self.prog)
+        else:
+            # NOTE: currently, only the gaussian backend supports shots > 1
+            results = self.eng.run(self.prog, shots=self.shots)
+
+        self.state = results.state
+        self.samples = results.samples
+
+    def probability(self, wires=None):
+        wires = wires or self.wires
+        wires = Wires(wires)
+
+        if self.analytic:
+            return super().probability(wires=wires)
+
+        N = len(wires)
+        samples = np.take(self.samples, self.wires.indices(wires), axis=1)
+
+        probs = all_fock_probs_pnr(samples)
+        ind = np.indices([self.cutoff] * N).reshape(N, -1).T
+
+        probs = OrderedDict((tuple(i), probs[tuple(i)]) for i in ind)
+        return probs
+
+    def jacobian(self, operations, observables, variable_deps):
+        """Calculate the Jacobian of the device.
+
+        Args:
+            operations (list[~pennylane.operation.Operation]): operations to be applied to the
+                device
+            observables (list[~pennylane.operation.Operation]): observables to be measured
+            variable_deps (dict[int, ParameterDependency]): reference dictionary
+                mapping free parameter values to the operations that
+                depend on them
+
+        Returns:
+            array[float]: Jacobian matrix of size (``len(probs)``, ``num_wires``)
+        """
+        self.reset()
+        obs_all_wires = qml.Identity(wires=self.wires)
+        obs_all_wires.return_type = Probability
+        prob = self.execute(operations, [obs_all_wires], parameters=variable_deps)[0]
+        requested_wires = observables[0].wires
+
+        jac = self._jacobian_all_wires(prob)
+
+        if requested_wires == self.wires:
+            return jac
+
+        jac = jac.reshape([self.cutoff] * self.num_wires + [self.num_wires])
+
+        all_indices = set(range(self.num_wires))
+        requested_indices = set(self.wires.indices(requested_wires))
+        trace_over_indices = all_indices - requested_indices
+
+        traced_jac = np.sum(jac, axis=tuple(trace_over_indices))
+        traced_jac = traced_jac.reshape(-1, self.num_wires)
```

Comment:
What's happening here? Are you computing the jacobian over all wires, and then summing over rows/cols representing wires not measured?

Reply:
That's right @josh146! I also added some comments to help follow along @ixfoduap